File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int n = s.size();
        int k = p.size();

        if (n < k) return {};
        
        vector<int> result;

        unordered_map<char, int> sCount, pCount;
        for (char c : p) pCount[c]++;

        int left = 0;
        for (int right = 0; right < n; right++) {
            // Expand the Window
            sCount[s[right]]++;
            // Check the window size
            if (right - left + 1 == k) {
                if (sCount == pCount) {
                    result.push_back(left);
                }
                // Shrink the window
                sCount[s[left]]--;
                if (sCount[s[left]] == 0) {
                    sCount.erase(s[left]);
                }
                left++;
            }
        }
        // Return the result
        return result;
    }
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def allZero(self, freq: dict) -> bool:
        for i in freq.values():
            if i != 0:
                return False
        return True

    def findAnagrams(self, s: str, p: str) -> List[int]:
        result = list()

        # Count Frequency
        freq = {}
        for i in p:
            freq[i] = freq.get(i, 0) + 1
        
        # Sliding Window
        i, j = 0, 0
        k = len(p)
        while j < len(s):
            # Decrease the freq count if there is an element in the array
            if s[j] in freq:
                freq[s[j]] -= 1
            
            if j - i + 1 == k:
                if self.allZero(freq):
                    result.append(i)
                
                if s[i] in freq:
                    freq[s[i]] += 1
                
                i += 1
            
            j += 1

        return result